httpbakery/form/form.go

```diff
@@ -0,0 +1,151 @@
+// Package form enables interactive login without using a web browser.
+package form
+
+import (
+	"net/http"
+	"net/url"
+
+	"github.com/juju/httprequest"
+	"gopkg.in/errgo.v1"
+	"gopkg.in/juju/environschema.v1"
+
+	"gopkg.in/macaroon-bakery.v1/httpbakery"
+)
+
+/*
+PROTOCOL
+
+A form login works as follows:
+
+	   Client                            Login Service
+	      |                                    |
+	      | GET visitURL with                  |
+	      | "Accept: application/json"         |
+	      |----------------------------------->|
+	      |                                    |
+	      |   Login Methods (including "form") |
+	      |<-----------------------------------|
+	      |                                    |
+	      | GET "form" URL                     |
+	      |----------------------------------->|
+	      |                                    |
+	      |                  Schema definition |
+	      |<-----------------------------------|
+	      |                                    |
+	+-------------+                            |
+	|   Client    |                            |
+	| Interaction |                            |
+	+-------------+                            |
+	      |                                    |
+	      | POST data to "form" URL            |
+	      |----------------------------------->|
+	      |                                    |
+	      |                Form login response |
+	      |<-----------------------------------|
+	      |                                    |
+
+The schema is provided as a environschema.Fileds object. It is the
+client's responsibility to interpret the schema and present it to the
+user.
+*/
+
+// Objects implementing the FormProcessor interface interpret schema and
+// interact with the user. When the interaction is complete the returned
+// value should conform to the provided schema.
+type Processor interface {
+	ProcessForm(schema environschema.Fields) (interface{}, error)
+}
+
+// SetUpAuth configures form authentication on c. The VisitWebPage field
+// in c will be set to a function that will attempt form based
+// authentication using h to perform the interaction with the user.
+//
+// The VisitWebPage function downloads the schema from the specified
+// server and calls h.Handle. The map returned by h.Handle should match
+// the schema specified, but VisitWebPage will not verify this before
+// sending to the server. Any errors returned by h.Handle will not have
+// their cause masked.
+func SetUpAuth(c *httpbakery.Client, p Processor) {
```

Comment:
Hm. If we go this route in the future, we need to make it "predictable" and well defined. There are environments, where web is not an option.